src/os/bluestore/BlueStore.cc

```diff
@@ -12399,6 +12399,40 @@ void BlueStore::_zoned_clean_zone(uint64_t zone_num) {
   dout(10) << __func__ << " cleaning zone " << zone_num << dendl;
   // TODO: (1) copy live objects from zone_num to a new zone, (2) issue a RESET
   // ZONE operation to the device for the corresponding zone.
+
+  zone_state_t zone_state;
+  std::string pfx = _zoned_get_prefix(zone_num * bdev->get_zone_size());
+  KeyValueDB::Iterator it = db->get_iterator(pfx, KeyValueDB::ITERATOR_NOCACHE);
+
+  while (it->valid())
+  {
+    std::string k = it->key();
+    bufferlist bl = it->value();
+    auto p = bl.cbegin();
+    int64_t offset;
+    ::decode(offset, p);
+    dout(40) << __func__ << " Copying object with key (zone_num + oid): " << k << " and offset: " << offset << dendl;
+    ghobject_t oid;
+    int r = get_key_object(k, &oid);
+    CollectionRef c = _get_collection(coll_t::meta());
+    OnodeRef o = c->get_onode(oid, false);
+    o->extent_map.fault_range(db, 0, o->onode.size);
+    ceph_assert(offset == o->zoned_get_ondisk_starting_offset());
+    //Should I use _read here and the flag = CEPH_OSD_OP_FLAG_FADVISE_NOCACHE
+    r = _do_read(c.get(), o, 0, o->onode.size, bl, 0);
```

Comment:
There are some concerns about the code starting from this line:
1) Object isn't locked at this point - hence possible races with other users.
2) reading and writing are not within an atomic transaction.
3) Reading in a single shot is doubtful (from RAM usage point of view) when dealing with huge objects. Suggest to perform using limited block sizes. Preferably by using underlying continuous physical extents  which are of limited size by their nature and benefit from the lack of fragmentation. Maybe multiple pextents to be processed in a single shot (atomic transaction) shot.
4) One doesn't need to make any object 'clones'. This should be just an atomic move operation on object's data (or a set of atomic operations). I'd dream about it's implementation as a standalone method enumerating onode's extents (or blobs?) and hence effectively determining blocks to process.
5) Please remind me whether we pin object exclusively to a specific zone? Can it have pextents from multiple zones? If so - we might need to handle that here...